File: algorithms/0206.reverse_linked_list/solution.ts
class ListNode {
  val: number
  next: ListNode | null

  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val
    this.next = next === undefined ? null : next
  }
}

// 206. Reverse Linked List (Easy)
// https://leetcode.com/problems/reverse-linked-list/

// Iterative solution
const reverseList = (head: ListNode | null): ListNode | null => {
  let prev: ListNode | null = null
  let curr: ListNode | null = head

  while (curr) {
    const next = curr.next
    curr.next = prev
    prev = curr
    curr = next
  }

  return prev
}

File: algorithms/0206.reverse_linked_list/solution.go
package problems

type ListNode struct {
	Val  int
	Next *ListNode
}

// 206. Reverse Linked List (Easy)
// https://leetcode.com/problems/reverse-linked-list/

// Iterative solution
func ReverseList(head *ListNode) *ListNode {
	var prev *ListNode = nil
	curr := head

	for curr != nil {
		next := curr.Next
		curr.Next = prev
		prev = curr
		curr = next
	}

	return prev
}
